### ISSUE TITLE
reveal -P doesn't work on macos with coreutils's stat in PATH

### ISSUE THREAD

**Issue by brandemix** [sobolevn/git-secret#572]
What are the steps to reproduce this issue?
-------------------------------------------
1. git secret reveal -FPf

What happens?
-------------
stat: cannot read file system information for '%04OLp': No such file or directory

What were you expecting to happen?
----------------------------------
No issues

Any logs, error output, etc?
----------------------------
stat: cannot read file system information for '%04OLp': No such file or directory

Any other comments?
-------------------
Looks like macos 10.15 Catalina version of stat doesn't use -f (--format) anymore, but -c now.
In the function '__get_octal_perms_osx' in src/utils/_git_secret_tools_osx.sh, the command should be
`stat -c "%04a" "$filename"` when on macos 10.15 Catalina

What versions of software are you using?
----------------------------------------
**Operating system:** 
Darwin 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64 i386 MacBookPro16,1 Darwin

**`git-secret` path:** 
/usr/local/bin/git-secret

**`git-secret` version:**
0.3.2

**`git` version:**
git version 2.21.1 (Apple Git-122.3)

**Shell type and version:**
zsh 5.7.1 (x86_64-apple-darwin19.0)

**`gpg` version:**
gpg (GnuPG) 2.2.19
libgcrypt 1.8.5

**Comment by joshrabinowitz** [comment]
Hello @brandemix and thank you for this issue report. 

The command you suggested, `stat -c "%04a" "$filename"` does not work on Mojave/OSX 10.14, I get `stat: illegal option -- c`.

Are you aware of an incantation for `stat` (or some other command) that would return the octal permissions on both Catalina and previous versions of OSX? 

(I'd prefer not to have to add a test for which version of OSX we're running. Also Travis-ci doesn't support testing on Catalina yet so we can't test this there, and I don't have a Catalina system set up yet to test this on.)

**Comment by joshrabinowitz** [comment]
See also: https://travis-ci.community/t/macos-catalina-build-environment/5608

**Comment by brandemix** [comment]
Might have helped to remember I installed gnu coreutils. 

@joshrabinowitz I guess my description of the error is not correct. `stat -f "%04OLp" <file>` works on macos 10.15 Catalina if I use the right version of `stat`. 
If a mac user is using the gnu coreutils package, git-secret doesn't account for that.

**Comment by joshrabinowitz** [comment]
@brandemix thank you for the follow up.

Firstly, I tried to replicate the issue on 10.14 Mojave using `homebrew` to install `coreutils`, but on my system that did not add a new version of `stat` to my path. How did you install gnu coreutils?

(Also, we've come across this sort of underlying issue before, where brew installed a command replacement which appears in the PATH before a system one -- see #485, where the issue was with `mktemp`.  The solution we used then is to use the full path to the command `/usr/bin/mktemp` instead of using PATH to find `mktemp`. But before we work up a solution I'd like to understand the case better first.)

**Comment by brandemix** [comment]
@joshrabinowitz Using the full path might work here too.

I used homebrew to install too, but I added this to my profile to tell my system to use the gnu utils before system utils. 
`export PATH="$(brew --prefix coreutils)/libexec/gnubin:$PATH"`

**Comment by joshrabinowitz** [comment]
Hi @brandemix 

I think that because you're putting `$(brew --prefix coreutils)/libexec/gnubin` in your path before the system bin directies, you're getting the version of `stat` that is there instead of the normal OSX version.

If you just add `/usr/local/bin` to your path as homebrew suggests, you'll be able to access the `gnu` version of `stat` as `gstat` (on my system, `which gstat` yields `/usr/local/bin/gstat`, and `stat` will still refer to the OSX supplied version, which is what `git-secret` expects.)

Since I think you have to specifically add `$(brew --prefix coreutils)/libexec/gnubin` to your path to get the gnu version of stat, I think this is Not A Bug as far as git-secret goes.

As mentioned at [this stack exchange page](https://superuser.com/questions/324616/how-should-i-set-the-path-variable-on-my-mac-so-the-hombrew-installed-tools-are) , 

````
As I understand it, brew doesn't put anything in /usr/local/bin 
that collides (has the same name as) an Apple distributed 
executable. Therefore, having /usr/local/bin in the path before 
/bin and /usr/bin shouldn't be an issue, because there should be 
no name collisions.
```` 

So in summary, I think the fix for this is _not_ to put the directory with gnu's `stat` in your PATH before the /bin and /usr/bin, and instead to put `/usr/local/bin` in your path (likey before /bin and /usr/bin, as the homebrew installer does) to use `gstat` (which will be in /usr/local/bin) if you need the gnu coreutils version of `stat`.

Let me know if you agree and/or have more issues,

**Comment by brandemix** [comment]
@joshrabinowitz Yeah, I was playing around with some shell voodoo, ended up not really needing gnu utils. I changed back to the system utils, and am not having issues anymore. 

I would agree this isn't really a bug in git-secret, and can probably be closed. Thanks for taking a look into this though, and hopefully anyone else who followed some guide to "Install and Use GNU command line tools on macos" can reference this issue.
